zmq.hpp

```diff
@@ -26,22 +26,37 @@
 #ifndef __ZMQ_HPP_INCLUDED__
 #define __ZMQ_HPP_INCLUDED__
 
-#if (__cplusplus >= 201402L)
+// macros defined if has a specific standard or greater
+#if (defined(__cplusplus) && __cplusplus >= 201103L) || (defined(_MSC_VER) && _MSC_VER >= 1900)
+    #define ZMQ_CPP11
+#endif
+#if (defined(__cplusplus) && __cplusplus >= 201402L) || (defined(_HAS_CXX14) && _HAS_CXX14 == 1)
+    #define ZMQ_CPP14
+#endif
+#if (defined(__cplusplus) && __cplusplus >= 201703L) || (defined(_HAS_CXX17) && _HAS_CXX17 == 1)
+    #define ZMQ_CPP17
+#endif
+
+#if defined(ZMQ_CPP14)
 #define ZMQ_DEPRECATED(msg) [[deprecated(msg)]]
 #elif defined(_MSC_VER)
 #define ZMQ_DEPRECATED(msg) __declspec(deprecated(msg))
 #elif defined(__GNUC__)
 #define ZMQ_DEPRECATED(msg) __attribute__((deprecated(msg)))
 #endif
 
-#if (__cplusplus >= 201103L) || (defined(_MSC_VER) && (_MSC_VER >= 1900))
-#define ZMQ_CPP11
+#if defined(ZMQ_CPP17)
+#define ZMQ_NODISCARD [[nodiscard]]
```

Comment:
This adds a definition of ZMQ_NODISCARD, but it this macro is not used anywhere. Are you working on a adding uses of this to zmq.hpp?

Reply:
Yes, `bool empty()` for message_t would be useful, where nodiscard would be appropriate.

Reply:
Sure. I think there are many other methods as well where nodiscard would make sense.

My question was if you intend to add these attributes in another PR soon, so we don't have a state with this unused definition around for long.

Reply:
Yes, I could do that within a couple of days.

Reply:
Great, I will merge this PR now.